pkg/informer/informer.go

```diff
@@ -643,3 +701,123 @@ func (d *DynamicDiscoverySharedInformerFactory) Unsubscribe(id string) {
 
 	delete(d.subscribers, id)
 }
+
+type discoveryData struct {
+	apiGroupResources []*restmapper.APIGroupResources
+	apiResourceList   []*metav1.APIResourceList
+}
+
+var _ discovery.ServerResourcesInterface = &DynamicDiscoverySharedInformerFactory{}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	for _, apiResourceList := range d.discoveryData.apiResourceList {
+		if apiResourceList.GroupVersion == groupVersion {
+			return apiResourceList.DeepCopy(), nil
+		}
+	}
+
+	// ignore 403 or 404 error to be compatible with a v1.0 server.
+	if groupVersion == "v1" {
+		return &metav1.APIResourceList{GroupVersion: groupVersion}, nil
+	}
+
+	return nil, errors.NewNotFound(schema.GroupResource{Group: groupVersion}, "")
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	retGroups := make([]*metav1.APIGroup, len(d.discoveryData.apiGroupResources))
+	for i, apiGroupResources := range d.discoveryData.apiGroupResources {
+		retGroups[i] = apiGroupResources.Group.DeepCopy()
+	}
+
+	retResourceList := make([]*metav1.APIResourceList, len(d.discoveryData.apiResourceList))
+	for i, apiResourceList := range d.discoveryData.apiResourceList {
+		retResourceList[i] = apiResourceList.DeepCopy()
+	}
+
+	return retGroups, retResourceList, nil
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerPreferredResources() ([]*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	ret := make([]*metav1.APIResourceList, len(d.discoveryData.apiResourceList))
+	for i, apiResourceList := range d.discoveryData.apiResourceList {
+		ret[i] = apiResourceList.DeepCopy()
+	}
+
+	return ret, nil
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerPreferredNamespacedResources() ([]*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	ret := make([]*metav1.APIResourceList, len(d.discoveryData.apiResourceList))
+	for i, apiResourceList := range d.discoveryData.apiResourceList {
+		namespacedResources := &metav1.APIResourceList{GroupVersion: apiResourceList.GroupVersion}
+		for _, resource := range apiResourceList.APIResources {
+			if resource.Namespaced {
+				namespacedResources.APIResources = append(namespacedResources.APIResources, resource)
+			}
+		}
+		ret[i] = namespacedResources
+	}
+
+	return ret, nil
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) RESTMapper() meta.ResettableRESTMapper {
+	return &d.restMapper
+}
+
+func newRESTMapper(fn func() (meta.RESTMapper, error)) restMapper {
+	return restMapper{
+		delegate: meta.NewLazyRESTMapperLoader(fn),
+	}
+}
+
+type restMapper struct {
+	delegate meta.RESTMapper
```

Comment:
Why not just embed this, to save yourself the methods?

Reply:
Done. This is SO much better 😃!